### ISSUE TITLE
Lab4 (lab 10) peer review - Davide Vitabile s330509

### ISSUE THREAD

**Issue by Vitabile** [HerryTheBest/Polito---Computational-Intelligence#9]
Hi Luca!

First of all I want to congratulate you, your work is great, the code is clearly written and easy to read with good comments and explanations.

I have some notes and tips regarding your implementation to improve your results.
First, you need to change the way you make a move in test time. Because it is a good method to use an exploration rate to increase the number of states known during training by trying random moves. But during the test the agent has to play as good as possible. So you need to remove this check and play random only when the agent doesn't know the state.  I tested your code with this change and with only 100k games in training I got an agent that never loses as first!

<img width="523" alt="image" src="https://github.com/HerryTheBest/Polito---Computational-Intelligence/assets/107575136/6678061d-a5df-491b-ba8f-11af0b4674d8">

Then, I appreciate the idea of taking into account the difference between playing as first or second. But your results are not as good as second. In my opinion, the problem is that we are in a game with two players, one against the other, so the "next_state" is not a real next_state where our agent could play. 
So: 
``` python 
optimal_next_reward = max(self.learning_dictionary[hashable_next_state][1])
```
for the update of q-table does not work.
To improve it, I advise you to try my implementation, I trained a single agent able to play as first and as second randomly playing in both the situation. But, I modified the update of the q-table adding a minus.
So: 
``` python
optimal_next_reward = -max(self.learning_dictionary[hashable_next_state][1]) 
```

In this way I'm trying to reward a state move that leads me to a "negative" adversarial state. (Because if the next state has a high expected reward, the minus will decrease our expected reward in the current state, and vice versa).
With this change, I got an agent able to not lose 100% of the time as first and as second! To test it, you need to change your implementation of training to get a single agent that must maximize the reward both as first and as second, and add the minus in the formula.
For other information you can see my code. This is a screen by my code of my results:
<img width="221" alt="image" src="https://github.com/HerryTheBest/Polito---Computational-Intelligence/assets/107575136/b71a18db-a074-48dc-8c73-a88a3715bdec">

In conclusion, good work and good luck on the exam!
